[scan] reorganize order of operations for simulator management in scan action.

jaleksynas opened the pull request:
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
Makes simulator management options behave as expected when used together. Suggested fix for: #14717 

### Description
- Move all simulator management operations to the start. 
- Add option and method to force quit simulators
- Add option conflicts so they reflect the `elsif` usage

Ran all automated tests. validated manually that the messaging from the issue did not occur.

jaleksynas committed:
Reorganize order of operations for simulator mangement in scan action.

janpio commented on scan/lib/scan/options.rb:
```diff
@@ -70,19 +70,30 @@ def self.available_options
                                      default_value: false,
-                                     type: Boolean),
-
-        # reinstall app
```
this structuring comment is still valid, right? please keep this.

jaleksynas replied:
I think it is a group of options (rather than simply restating the name of a single option).
I added a ` simulator management` structural comment for the related three options.

janpio replied:
Yes, before there were two different "group" comments, one just repeating the obvious as I couldn't find a better label. What you did now think is fine and matches the intention behind those options just fine.

janpio commented on scan/lib/scan/options.rb:
```diff
@@ -70,19 +70,30 @@ def self.available_options
                                      type: Boolean,
                                      optional: true),
 
-        # reset simulator
```
this structuring comment is not valid any more as `force_quit_simulator` was added and "belongs" to `reset_simulator` thematically, correct? If so, what would be a new description of what these to do? "before simulator usage" maybe?

janpio commented:
Good catch re the conflicting options! 

Left two very nitpicky comments re the comment.

jaleksynas commented on scan/lib/scan/runner.rb:
```diff
@@ -23,12 +23,28 @@ def run
     def test_app
+      force_quit_simulator_processes if Scan.config[:force_quit_simulator]
+
+      if Scan.config[:reset_simulator]
```
Should simulators be forced stopped here to avoid spurious errors?

jaleksynas commented on scan/lib/scan/runner.rb:
```diff
@@ -163,10 +166,15 @@ def copy_simulator_logs
     end
+
+    def force_quit_simulator_processes
+      UI.message("Killing all running simulators")
```
Should this change to `other_action.sh({log:false])`  Then there is no log impacts unless debug mode is enabled.

jaleksynas commented:
I included some of the details mentioned in #14721 as well.

janpio commented:
(No need to force push and squash commits, we squash on merge anyway so it is easier to follow PRs with incremental commits)

raymondhoagland commented:
From the changes/description the behavior it looks like we're going to is

## Options
| Name                 | Behavior                                   | Conflicts                            | Workflow                                        |
|----------------------|--------------------------------------------|--------------------------------------|-------------------------------------------------|
| force_quit_simulator | Kill all running simulators                | reinstall_app                        | This happens before any other mgmt              |
| reset_simulator      | Reset a set of emulators (set by :devices) | reinstall_app                        | After killing any running simulators (optional) |
| prelaunch_simulator  | Launch a simulator (legacy behavior?)      | none                                 | After killing/resetting simulators (optional)   |
| reinstall_app        | Reinstall the app on a running simulator   | force_quit_simulator | After prelaunch (optional)                      |

Based on that chart, behavior I'd expect to see falls into a couple categories as follows:


1) Emulator was already running

| Option               | Value | Behavior                                        |
|----------------------|-------|-------------------------------------------------|
| force_quit_simulator | True  | Kill the running simulator                      |
| force_quit_simulator | False | no-op                                           |
| reset_simulator      | True  | Error (rescued) unless force_quit first, otherwise reset  |
| reset_simulator      | False | no-op                                           |
| prelaunch_simulator  | True  | Error unless force_quit first, otherwise launch |
| prelaunch_simulator  | False | no-op                                           |
| reinstall_app        | True  | Error if force_quit used (and not prelaunch), otherwise reinstall   |
| reinstall_app        | False | no-op                                           |

2) Emulator not running

| Option               | Value | Behavior                                         |
|----------------------|-------|--------------------------------------------------|
| force_quit_simulator | True  | no-op (kill)                                     |
| force_quit_simulator | False | no-op                                            |
| reset_simulator      | True  | Reset the simulators                             |
| reset_simulator      | False | no-op                                            |
| prelaunch_simulator  | True  | Launch the simulator                             |
| prelaunch_simulator  | False | no-op                                            |
| reinstall_app        | True  | Error unless prelaunch used, otherwise reinstall |
| reinstall_app        | False | no-op                                            |

The conflict between force_quit_simulator and reinstall_app is true _unless_ prelaunch is also used to start the simulator.  For example, if the user specifies all four options, the simulator will be quit, reset, launched, and then the app will be reinstalled, which should still work -- so there isn't necessarily a conflict.  Do we err on the side of expecting people to follow a "preferred" flow of just calling reinstall with an _existing_ simulator?  Or do we sidestep this and allow builds to fail if the user follows an invalid workflow e.g. force_quit & reinstall_app only?

jaleksynas commented:
@raymondhoagland I believe that less opinionated is better, I'd be glad to remove the conflict blocks since there are combinations that allow for proper functioning.  

@joshdholtz @janpio  Any thoughts or direction would be appreciated so this PR can progress.

janpio reviewed (approved):
To be honest this is a bit over my head. What @raymondhoagland summarized looks good to me, and you obviously spent a lot more time thinking this through than I ever could while reviewing. Do you see any further need to change things here, removing the conflict stuff again?

jaleksynas committed:
Remove conflict blocks allow users control

The option combinations for simulator controls are complex. Allow users to decide how best to use them.

raymondhoagland commented:
That last change addresses my concerns :) lgtm!

joshdholtz reviewed (approved):
This is super clean 💯 Thanks for putting the time in to add this and thanks to @janpio for all the reviews! ❤️
